sotodlib/io/bookbinder2.py

```diff
@@ -0,0 +1,1118 @@
+                telescope is slewing.
+            - azimuth_velocity_mean / azimuth_velocity_std: (float / float)
+                Mean and standard deviation of the az velocity (deg / sec)
+            - elevation_velocity_mean / elevaction_velocity_std: (float / float)
+                Mean and standard deviation of the el velocity (deg / sec)
+            
+        Params
+        ----------
+        frame : G3Frame
+            Frame to add data to
+        t0 : float
+            Start time of frame (unix time)
+        t1 : float
+            Stop time of frame (unix time)
+        """
+        bl = self.blocks.get('ACU_summary_output')
+        if bl.data:
+            az_mode = bl.data["Azimuth_mode"]
+            acu_summary_times = bl.times
+            m = (t0 < acu_summary_times) & (acu_summary_times <= t1)
+            if not np.any(m):
+                frame['azimuth_mode'] = 'None'
+            elif 'ProgramTrack' in az_mode[m]:
+                frame['azimuth_mode'] = 'ProgramTrack'  # Scanning
+            else:
+                frame['azimuth_mode'] = 'Preset'  # "Slewing"
+
+        bl = self.blocks['ACU_broadcast']
+        if bl.data:
+            az = bl.data["Corrected_Azimuth"]
+            el = bl.data["Corrected_Elevation"]
+            m = (t0 < bl.times) & (bl.times <= t1)
+            dt = np.diff(bl.times[m]).mean()
+            az_vel = np.diff(az[m]) / dt
+            el_vel = np.diff(el[m]) / dt
+
+            frame['azimuth_velocity_mean'] = np.mean(az_vel)
+            frame['azimuth_velocity_stdev'] = np.std(az_vel)
+            frame['elevation_velocity_mean'] = np.mean(el_vel)
+            frame['elevation_velocity_stdev'] = np.std(el_vel)
+
+        
+class SmurfStreamProcessor:
+    def __init__(self, obs_id, files, book_id, readout_ids, log=None):
+        self.files = files
+        self.obs_id = obs_id
+        self.stream_id = None
+        self.times = None
+        self.frame_idxs = None
+        self.nchans = None
+        self.nframes = None
+        self.bias_names = None
+        self.primary_names = None
+        self.timing_paradigm = None
+        self.session_id = None
+        self.slow_primary = None
+        self.sostream_version = None
+        self.readout_ids = readout_ids
+        self.out_files = []
+        self.book_id = book_id
+
+        if log is None:
+            self.log = logging.getLogger('bookbinder')
+        else:
+            self.log = log
+
+    def preprocess(self):
+        """
+        Extracts file times, nchans, and nnframes from file list
+        """
+        if self.times is not None:  # Already preprocessed
+            return
+
+        self.log.info(f"Preprocessing smurf obsid {self.obs_id}")
+
+        self.nframes = 0
+        ts = []
+        frame_idxs = []
+        frame_idx = 0
+        for frame in get_frame_iter(self.files):
+            if frame.type != core.G3FrameType.Scan:
+                continue
+
+
```

Comment:
Is the extra empty line meant to be a comment? (Can we add one here, something like `# Populate attributes`?)